typing/rec_check.ml

```diff
@@ -18,145 +18,161 @@ open Types
+       a variable in an expression in which it does not occur. *)
+
+  (* Returns the most conservative mode of the two arguments.
+
+     Dereference < Return < Guard < Delay < Unused
+
+     In judgments we write (m + m') for (join m m').
+  *)
+  let join m m' =
+    match m, m' with
+    | Dereference, _
+    | _, Dereference -> Dereference
+    | Return, _
+    | _, Return -> Return
+    | Guard, _
+    | _, Guard -> Guard
+    | Delay, _
+    | _, Delay -> Delay
+    | _ -> Unused
+
+  (* If x is used with the mode m in e[x], and e[x] is used with mode
+     m' in e'[e[x]], then x is used with mode m'[m] in e'[e[x]].
+
+     Return is neutral for composition: m[Return] = m = Return[m].
+  *)
+  let compos m' m = match m', m with
+    | Unused, _ | _, Unused -> Unused
+    | Dereference, _ -> Dereference
+    | Delay, _ -> Delay
+    | Guard, Return -> Guard
+    | Guard, ((Dereference | Guard | Delay) as m) -> m
+    | Return, Return -> Return
+    | Return, ((Dereference | Guard | Delay) as m) -> m
 end
 
-open Rec_context
+type mode = Mode.t = Dereference | Return | Guard | Delay | Unused
+
+module Env :
+sig
+  type t
+
+  val single : Ident.t -> Mode.t -> t
+  (** Create an environment with a single identifier used with a given mode.
+  *)
+
+  val empty : t
+  (** An environment with no used identifiers. *)
+
+  val find : Ident.t -> t -> Mode.t
+  (** Find the mode of an indentifier in an environment.  The default mode is
+      Unused. *)
+
+  val unguarded : t -> Ident.t list -> Ident.t list
+  (** unguarded e l: the list of all identifiers in l that are dereferenced or
+      returned in the environment e. *)
+
+  val dependent : t -> Ident.t list -> Ident.t list
+  (** unguarded e l: the list of all identifiers in e that are used in e. *)
+
+  val join : t -> t -> t
+  val join_list : t list -> t
+  (** Environments can be joined pointwise (variable per variable) *)
+
+  val remove : Ident.t -> t -> t
+  (** Remove an identifier from an environment. *)
+
+  val take: Ident.t -> t -> Mode.t * t
+  (** Remove an identifier from an environment, and return its mode *)
+
+  val remove_list : Ident.t list -> t -> t
+  (** Remove all the identifiers of a list from an environment. *)
+end = struct
+  module M = Map.Make(Ident)
+
+  (** A "t" maps each rec-bound variable to an access status *)
+  type t = Mode.t M.t
+
+  let find (id: Ident.t) (tbl: t) =
+    try M.find id tbl with Not_found -> Unused
+
+  let empty = M.empty
 
-let build_unguarded_env : Ident.t list -> Env.env = fun idlist ->
-  List.fold_left
-    (fun env id -> Ident.add id (Use.single id Unguarded) env)
-    Env.empty
-    idlist
+  let join (x: t) (y: t) =
+    M.fold
+      (fun (id: Ident.t) (v: Mode.t) (tbl: t) ->
+         let v' = find id tbl in
+         M.add id (Mode.join v v') tbl)
+      x y
+
+  let join_list li = List.fold_left join empty li
+
+  let single id mode = M.add id mode empty
+
+  let unguarded env li =
```

Comment:
Here and in `join` you define an order in the comments, and use it implicitly in the code in various places (`join`, `unguarded`, `dependent`).

Have you considered instead defining the order directly in the code (in fact, the order given by polymorphic `<` would do) , then making use of it explicitly?  For example, you could then define `unguarded` like this:

```ocaml
let unguarded env = List.filter (fun id -> lt (find id env) Guard)
```

which is clearer, since it makes use of the semantic notion of mode ordering rather than listing all the constructors.

Reply:
I defined a `rank : mode -> int` function to use binary ordering operators without feeling dirty about it. The code is much nicer, thanks!

I decided to change the order I used in this comment: instead of `Dereference < ... < Ignore`, the new version I just pushed uses `Ignore < ... < Dereference`¹. I think that it makes much sense (although of course the two orders are dual): we are talking about how a context "uses" a subterm, or a term uses a variable, and intuitively Ignore "uses less". 

¹: It would be fun to define this new order without changing the order of constructors in the datatype declaration, but I'm not this kind of mean person so I did change the constructor order.